src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,210 @@
                 let buckets = slice::from_raw_parts(buckets_ptr, nbuckets as usize);
-                let hashvalues_ptr = buckets_ptr.add(nbuckets as usize);
-                let hashvalues = slice::from_raw_parts(hashvalues_ptr, total_dynsyms - symindex);
-                let bloomwords = slice::from_raw_parts(bloomwords_ptr, maskwords);
-                GnuHash {
-                    nbuckets,
+                let chains = slice::from_raw_parts(chains_ptr, total_dynsyms - symindex as usize);
+                Ok(Self {
                     symindex,
                     shift2,
-                    maskbits: mem::size_of::<usize>() as u32,
-                    bloomwords,
-                    hashvalues,
+                    bloom_filter,
                     buckets,
-                    maskwords_bitmask: ((maskwords as i32) - 1) as u32,
-                    symtab,
-                }
+                    chains,
+                })
             }
 
-            #[inline(always)]
-            fn lookup(&self,
-                      symbol: &str,
-                      hash: u32,
-                      strtab: &Strtab)
-                      -> Option<&sym::Sym> {
-                let mut idx = self.buckets[(hash % self.nbuckets) as usize] as usize;
-                // println!("lookup idx = buckets[hash % nbuckets] = {}", idx);
-                if idx == 0 {
+            /// Locate the hash chain, and corresponding hash value element.
+            fn lookup<'b>(&self, symbol: &str, hash: u32, dynsyms: &'b [Sym], dynstrtab: &Strtab) -> Option<&'b Sym> {
+                const MASK_LOWEST_BIT: u32 = 0xffff_fffe;
+                let bucket = self.buckets[hash as usize % self.buckets.len()];
+
+                // Empty hash chain, symbol not present
+                if bucket < self.symindex {
                     return None;
                 }
-                let mut hash_idx = idx - self.symindex;
-                let hash = hash & !1;
-                // TODO: replace this with an iterator
-                loop {
-                    let symbol_ = &self.symtab[idx];
-                    let h2 = self.hashvalues[hash_idx];
-                    idx += 1;
-                    hash_idx += 1;
-                    let name = &strtab[symbol_.st_name as usize];
-                    // println!("{}: h2 0x{:x} resolves to: {}", i, h2, name);
-                    if hash == (h2 & !1) && name == symbol {
-                        // println!("lookup match for {} at: 0x{:x}", symbol, symbol_.st_value);
-                        return Some(symbol_);
+                // Walk the chain until the symbol is found or the chain is exhausted.
+                let chain_idx = bucket - self.symindex;
+                let hash = hash & MASK_LOWEST_BIT;
+                let chains = &self.chains[(chain_idx as usize)..];
+                let dynsyms = &dynsyms[(bucket as usize)..];
```

Comment:
If you want to better tolerate corrupt data instead of panicking, you could use `slice::get` instead (for both `chains` and `dynsyms`). Maybe this is what you wanted the dynsyms length for, but you don't need that separately because slices already have a length.